elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +305,321 @@ private Object fetchRelationship(Environment context) {
+        for(Map<String, Object> input : context.data.get()) {
+            upsertGraphWalker(input, this::updateRelationship, Optional.ofNullable(context.parentResource),
+                    context.requestScope, context.field.getName());
+        }
+
+        return upsertOps;
+    }
+
+    /**
+     * utility class to support storing a triplet on each node of the queue used while upserting
+     */
+    private class Triplet {
+        private List<Map<String, Object>> input;
+        private Optional<PersistentResource> parentResource;
+        private String fieldName;
+
+        /**
+         * class constructor
+         * @param input data input
+         * @param parentResource parent
+         * @param fieldName field name
+         */
+        private Triplet(List<Map<String, Object>> input, Optional<PersistentResource> parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * A function to handle upserting (update/create) objects.
+     */
+    @FunctionalInterface
+    private interface UpsertInterface<T> {
+        T execute(Map<String, Object> input, String id,
+                                   RequestScope requestScope, Class<?> entityClass,
+                                   String fieldName, Optional<PersistentResource> parent);
+    }
+
+    /**
+     * Forms the graph from data {@param input} and upsert's {@param function} all the nodes
+     * @param input input data
+     * @param function function to process nodes
+     * @param parent parent resource
+     * @param requestScope request scope
+     * @param fieldName graphql field name
+     * @return set of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertGraphWalker(Map<String, Object> input, UpsertInterface function,
+                                                      Optional<PersistentResource> parent, RequestScope requestScope,
+                                                      String fieldName) {
+        EntityDictionary dictionary = requestScope.getDictionary();
+        Queue<Triplet> toVisit = new ArrayDeque();
+        Set visited = new HashSet<Triplet>();
+        toVisit.add(new Triplet(new ArrayList<>(Collections.singletonList(input)), parent, fieldName));
+        Set<PersistentResource> toReturn = new HashSet<>();
+
+        while(!toVisit.isEmpty()) {
+            Triplet triplet = toVisit.remove();
+
+            if(visited.contains(triplet)) {
+                continue;
+            } else {
+                visited.add(triplet);
+            }
+            Class<?> entityClass = getEntityClass(triplet.parentResource, dictionary, triplet.fieldName);
+            String idFieldName = getIdFieldName(dictionary, entityClass);
+            Optional<String> id = getId(input, idFieldName);
+            if(!id.isPresent()) {
+                setId(input, idFieldName);
```

Comment:
The while loop should never directly reference the input parameter passed into the function.  It should purely operate from the triplet it pops from the queue.